lib/shelley/src/Cardano/Wallet/Shelley/Network/Blockfrost.hs

```diff
@@ -125,35 +152,166 @@ withNetworkLayer tr project k = k NetworkLayer
       where
         pollNodeTip :: (BlockHeader -> IO ()) -> IO ()
         pollNodeTip cb = forever $ do
-            runExceptT fetchLatestBlockHeader >>= \case
+            runBlockfrost BF.getLatestBlock & runExceptT >>= \case
                 Left err -> throwIO (BlockfrostException err)
                 Right header ->
                     bracketTracer (MsgWatcherUpdate header >$< tr) $ cb header
             threadDelay 2_000_000
 
-    fetchLatestBlockHeader :: ExceptT BlockfrostError IO BlockHeader
-    fetchLatestBlockHeader =
-        runBlockfrost BF.getLatestBlock >>= blockToBlockHeader
+    currentProtocolParameters :: IO ProtocolParameters
+    currentProtocolParameters =
+        runBlockfrost BF.getLatestEpochProtocolParams & runExceptT >>= \case
+            -- TODO: use cached value while retrying
+            Left err -> throwIO (BlockfrostException err)
+            Right params -> pure params
 
-    runBlockfrost :: BF.BlockfrostClientT IO a -> ExceptT BlockfrostError IO a
+    runBlockfrost ::
+        FromBlockfrost b w =>
+        BF.BlockfrostClientT IO b ->
+        ExceptT BlockfrostError IO w
     runBlockfrost =
-        ExceptT . (first ClientError <$>) . BF.runBlockfrostClientT project
-
-blockToBlockHeader ::
-    forall m. MonadError BlockfrostError m => BF.Block -> m BlockHeader
-blockToBlockHeader block@BF.Block{..} = do
-    slotNo <- case _blockSlot of
-        Just s -> pure $ SlotNo $ fromIntegral $ BF.unSlot s
-        Nothing -> throwError $ NoSlotError block
-    blockHeight <- case _blockHeight of
-        Just height -> pure $ Quantity $ fromIntegral height
-        Nothing -> throwError $ NoBlockHeight block
-    headerHash <- parseBlockHeader _blockHash
-    parentHeaderHash <- for _blockPreviousBlock parseBlockHeader
-    pure BlockHeader { slotNo, blockHeight, headerHash, parentHeaderHash }
-  where
-    parseBlockHeader :: BF.BlockHash -> m (Hash "BlockHeader")
-    parseBlockHeader blockHash =
-        case fromText (BF.unBlockHash blockHash) of
-            Right hash -> pure hash
-            Left tde -> throwError $ InvalidBlockHash blockHash tde
+        fromBlockfrost
+            <=< ExceptT
+            <<< (first ClientError <$>)
+            <<< BF.runBlockfrostClientT project
+
+class FromBlockfrost b w where
+    fromBlockfrost :: forall m. MonadError BlockfrostError m => b -> m w
```

Comment:
```suggestion
    fromBlockfrost :: b -> Either BlockfrostError w

fromBlockfrostM :: forall m. MonadError BlockfrostError m => a -> m w
fromBlockfrostM = liftEither . fromBlockfrost
```

My preference would be to use concrete types instead of general type classes, but I understand the convenience of not having to write `liftEither` explicitly at usage sites. The use of concrete types also makes it more clear that this is a pure function.